File: android/src/main/java/vn/hunghd/flutterdownloader/FlutterDownloaderInitializer.java
package vn.hunghd.flutterdownloader;

import android.content.ComponentName;
import android.content.ContentProvider;
import android.content.ContentValues;
import android.content.Context;
import android.content.pm.PackageManager;
import android.content.pm.ProviderInfo;
import android.database.Cursor;
import android.net.Uri;
import android.os.Bundle;
import android.util.Log;

import java.util.concurrent.Executors;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.work.Configuration;
import androidx.work.WorkManager;

public class FlutterDownloaderInitializer extends ContentProvider {
    private final String TAG = "DownloaderInitializer";
    private final int DEFAULT_MAX_CONCURRENT_TASKS = 3;

    @Override
    public boolean onCreate() {
        int maximumConcurrentTask = getMaxConcurrentTaskMetadata(getContext());
        WorkManager.initialize(getContext(), new Configuration.Builder()
                .setExecutor(Executors.newFixedThreadPool(maximumConcurrentTask))
                .build());
        return true;
    }

    @Nullable
    @Override
    public Cursor query(
            @NonNull Uri uri,
            @Nullable String[] strings,
            @Nullable String s,
            @Nullable String[] strings1,
            @Nullable String s1) {
        return null;
    }

    @Nullable
    @Override
    public String getType(@NonNull Uri uri) {
        return null;
    }

    @Nullable
    @Override
    public Uri insert(
            @NonNull Uri uri,
            @Nullable ContentValues contentValues) {
        return null;
    }

    @Override
    public int delete(
            @NonNull Uri uri,
            @Nullable String s,
            @Nullable String[] strings) {
        return 0;
    }

    @Override
    public int update(
            @NonNull Uri uri,
            @Nullable ContentValues contentValues,
            @Nullable String s,
            @Nullable String[] strings) {
        return 0;
    }

    private int getMaxConcurrentTaskMetadata(Context context) {
        try {
            ProviderInfo pi = context.getPackageManager().getProviderInfo(
                    new ComponentName(context, "vn.hunghd.flutterdownloader.FlutterDownloaderInitializer"),
                    PackageManager.GET_META_DATA);
            Bundle bundle = pi.metaData;
            int max = bundle.getInt("vn.hunghd.flutterdownloader.MAX_CONCURRENT_TASKS", DEFAULT_MAX_CONCURRENT_TASKS);
            Log.d(TAG, "MAX_CONCURRENT_TASKS = " + max);
            return max;
        } catch (PackageManager.NameNotFoundException e) {
            Log.e(TAG, "Failed to load meta-data, NameNotFound: " + e.getMessage());
        } catch (NullPointerException e) {
            Log.e(TAG, "Failed to load meta-data, NullPointer: " + e.getMessage());
        }
        return DEFAULT_MAX_CONCURRENT_TASKS;
    }
}

File: android/src/main/kotlin/vn/hunghd/flutterdownloader/FlutterDownloaderInitializer.kt
package vn.hunghd.flutterdownloader

import android.content.ComponentName
import android.content.ContentProvider
import android.content.ContentValues
import android.content.Context
import android.content.pm.PackageManager
import android.content.pm.PackageManager.NameNotFoundException
import android.net.Uri
import android.util.Log
import androidx.work.Configuration
import androidx.work.WorkManager
import java.util.concurrent.Executors

class FlutterDownloaderInitializer : ContentProvider() {
    companion object {
        private const val TAG = "DownloaderInitializer"
        private const val DEFAULT_MAX_CONCURRENT_TASKS = 3
    }

    override fun onCreate(): Boolean {
        val context = requireNotNull(this.context) { "Cannot find context from the provider." }
        val maximumConcurrentTask = getMaxConcurrentTaskMetadata(context)
        WorkManager.initialize(
            context,
            Configuration.Builder()
                .setExecutor(Executors.newFixedThreadPool(maximumConcurrentTask))
                .build()
        )
        return true
    }

    override fun query(uri: Uri, strings: Array<String>?, s: String?, strings1: Array<String>?, s1: String?): Nothing? = null

    override fun getType(uri: Uri): Nothing? = null

    override fun insert(uri: Uri, contentValues: ContentValues?): Uri? = null

    override fun delete(uri: Uri, s: String?, strings: Array<String>?) = 0

    override fun update(uri: Uri, contentValues: ContentValues?, s: String?, strings: Array<String>?) = 0

    private fun getMaxConcurrentTaskMetadata(context: Context): Int {
        try {
            val providerInfo = context.packageManager.getProviderInfo(
                ComponentName(context, "vn.hunghd.flutterdownloader.FlutterDownloaderInitializer"),
                PackageManager.GET_META_DATA
            )
            val bundle = providerInfo.metaData
            val max = bundle.getInt(
                "vn.hunghd.flutterdownloader.MAX_CONCURRENT_TASKS",
                DEFAULT_MAX_CONCURRENT_TASKS
            )
            Log.d(TAG, "MAX_CONCURRENT_TASKS = $max")
            return max
        } catch (e: NameNotFoundException) {
            Log.e(TAG, "Failed to load meta-data, NameNotFound: " + e.message)
        } catch (e: NullPointerException) {
            Log.e(TAG, "Failed to load meta-data, NullPointer: " + e.message)
        }
        return DEFAULT_MAX_CONCURRENT_TASKS
    }
}
